File: engine/static/render/src/2d/render_object_2d.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/render/2d/render_object_2d.hpp"

#include "argus/lowlevel/dirtiable.hpp"
#include "argus/lowlevel/memory.hpp"

#include "argus/render/2d/render_group_2d.hpp"
#include "argus/render/common/transform.hpp"
#include "internal/render/module_render.hpp"
#include "internal/render/pimpl/2d/render_object_2d.hpp"
#include "internal/render/pimpl/common/transform_2d.hpp"

#include <algorithm>
#include <string>
#include <vector>

namespace argus {
    // forward declarations
    class RenderPrim2D;

    class Scene2D;

    static PoolAllocator g_pimpl_pool(sizeof(pimpl_RenderObject2D));

    RenderObject2D::RenderObject2D(const RenderGroup2D &parent_group, const std::string &material,
            const std::vector<RenderPrim2D> &primitives, const Vector2f &anchor_point, const Vector2f &atlas_stride,
            uint32_t z_index, float light_opacity, const Transform2D &transform) :
            m_pimpl(&g_pimpl_pool.construct<pimpl_RenderObject2D>(g_render_handle_table.create_handle(this), parent_group,
                    material, primitives, anchor_point, atlas_stride, z_index, light_opacity, transform)) {
        m_pimpl->transform.set_version_ref(m_pimpl->version);
    }

    RenderObject2D::RenderObject2D(Handle handle, const RenderGroup2D &parent_group, const std::string &material,
            const std::vector<RenderPrim2D> &primitives, const Vector2f &anchor_point, const Vector2f &atlas_stride,
            uint32_t z_index, float light_opacity, const Transform2D &transform) :
            m_pimpl(&g_pimpl_pool.construct<pimpl_RenderObject2D>(handle, parent_group,
                    material, primitives, anchor_point, atlas_stride, z_index, light_opacity, transform)) {
        m_pimpl->transform.set_version_ref(m_pimpl->version);
        g_render_handle_table.update_handle(handle, this);
    }

    RenderObject2D::RenderObject2D(RenderObject2D &&rhs) noexcept:
            m_pimpl(rhs.m_pimpl) {
        rhs.m_pimpl = nullptr;
    }

    RenderObject2D::~RenderObject2D() {
        if (m_pimpl != nullptr) {
            g_render_handle_table.release_handle(m_pimpl->handle);
            g_pimpl_pool.destroy(m_pimpl);
        }
    }

    Handle RenderObject2D::get_handle(void) const {
        return m_pimpl->handle;
    }

    const Scene2D &RenderObject2D::get_scene(void) const {
        return m_pimpl->parent_group.get_scene();
    }

    const RenderGroup2D &RenderObject2D::get_parent(void) const {
        return m_pimpl->parent_group;
    }

    const std::string &RenderObject2D::get_material(void) const {
        return m_pimpl->material;
    }

    const std::vector<RenderPrim2D> &RenderObject2D::get_primitives(void) const {
        return m_pimpl->primitives;
    }

    const Vector2f &RenderObject2D::get_anchor_point(void) const {
        return m_pimpl->anchor_point;
    }

    const Vector2f &RenderObject2D::get_atlas_stride(void) const {
        return m_pimpl->atlas_stride;
    }

    uint32_t RenderObject2D::get_z_index(void) const {
        return m_pimpl->z_index;
    }

    float RenderObject2D::get_light_opacity(void) const {
        return m_pimpl->light_opacity.peek();
    }

    void RenderObject2D::set_light_opacity(float opacity) {
        m_pimpl->light_opacity = opacity;
    }

    ValueAndDirtyFlag<Vector2u> RenderObject2D::get_active_frame(void) const {
        return m_pimpl->active_frame.read();
    }

    void RenderObject2D::set_active_frame(const Vector2u &index) {
        m_pimpl->active_frame = index;
    }

    const Transform2D &RenderObject2D::peek_transform(void) const {
        return m_pimpl->transform;
    }

    Transform2D &RenderObject2D::get_transform(void) {
        return m_pimpl->transform;
    }

    void RenderObject2D::set_transform(const Transform2D &transform) {
        m_pimpl->transform = transform;
    }

    RenderObject2D &RenderObject2D::copy(RenderGroup2D &parent) {
        std::vector<RenderPrim2D> prims_copy;
        std::transform(m_pimpl->primitives.begin(), m_pimpl->primitives.end(), std::back_inserter(prims_copy),
                [](auto &v) { return RenderPrim2D(v); });
        auto new_handle = g_render_handle_table.copy_handle(m_pimpl->handle);
        auto &copy = *new RenderObject2D(new_handle, parent, m_pimpl->material, prims_copy, m_pimpl->anchor_point,
                m_pimpl->atlas_stride, m_pimpl->z_index, m_pimpl->light_opacity.peek(), m_pimpl->transform);
        copy.m_pimpl->active_frame = m_pimpl->active_frame;
        copy.m_pimpl->version = m_pimpl->version;

        g_render_handle_table.update_handle(m_pimpl->handle, copy);

        return copy;
    }
}

File: engine/auxiliary/render_rustabi/src/argus/render/twod/render_object_2d.rs
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::{mem, slice};

use lowlevel_rustabi::argus::lowlevel::{Handle, ValueAndDirtyFlag, Vector2f, Vector2u};
use lowlevel_rustabi::util::cstr_to_str;

use crate::argus::render::{RenderGroup2d, RenderPrimitive2d, Scene2d};
use crate::render_cabi::*;

pub struct RenderObject2d {
    handle: argus_render_object_2d_t,
}

impl RenderObject2d {
    pub(crate) fn of(handle: argus_render_object_2d_t) -> Self {
        Self { handle }
    }

    pub fn get_handle(&self) -> Handle {
        unsafe { argus_render_object_2d_get_handle(self.handle).into() }
    }

    pub fn get_scene(&self) -> Scene2d {
        unsafe { Scene2d::of(mem::transmute(argus_render_object_2d_get_scene(self.handle))) }
    }

    pub fn get_parent(&self) -> RenderGroup2d {
        unsafe { RenderGroup2d::of(mem::transmute(argus_render_object_2d_get_parent(self.handle))) }
    }

    pub fn get_material(&self) -> &str {
        unsafe { cstr_to_str(argus_render_object_2d_get_material(self.handle)) }
    }

    pub fn get_primitives(&self) -> Vec<RenderPrimitive2d> {
        unsafe {
            let count = argus_render_object_2d_get_primitives_count(self.handle);
            let mut buf: Vec<ArgusRenderPrimitive2d> = Vec::with_capacity(count);
            buf.resize(count, mem::zeroed());
            argus_render_object_2d_get_primitives(self.handle, buf.as_mut_ptr(), count);
            buf.into_iter()
                .map(|prim| RenderPrimitive2d {
                    vertices: slice::from_raw_parts(prim.vertices.cast(), prim.vertex_count)
                        .to_vec(),
                })
                .collect()
        }
    }

    pub fn get_anchor_point(&self) -> Vector2f {
        unsafe { argus_render_object_2d_get_anchor_point(self.handle).into() }
    }

    pub fn get_atlas_stride(&self) -> Vector2f {
        unsafe { argus_render_object_2d_get_atlas_stride(self.handle).into() }
    }

    pub fn get_z_index(&self) -> u32 {
        unsafe { argus_render_object_2d_get_z_index(self.handle) }
    }

    pub fn get_light_opacity(&self) -> f32 {
        unsafe { argus_render_object_2d_get_light_opacity(self.handle) }
    }

    pub fn set_light_opacity(&mut self, opacity: f32) {
        unsafe { argus_render_object_2d_set_light_opacity(self.handle, opacity); }
    }

    pub fn get_active_frame(&mut self) -> ValueAndDirtyFlag<Vector2u> {
        unsafe {
            let mut dirty = false;
            let val = argus_render_object_2d_get_active_frame(self.handle, &mut dirty);
            ValueAndDirtyFlag {
                value: val.into(),
                dirty,
            }
        }
    }

    pub fn set_active_frame(&mut self, frame: Vector2u) {
        unsafe { argus_render_object_2d_set_active_frame(self.handle, frame.into()); }
    }

    pub fn peek_transform(&self) -> ArgusTransform2d {
        unsafe { argus_render_object_2d_peek_transform(self.handle).into() }
    }

    pub fn get_transform(&mut self) -> ArgusTransform2d {
        unsafe { argus_render_object_2d_get_transform(self.handle).into() }
    }

    pub fn set_transform(&mut self, transform: ArgusTransform2d) {
        unsafe { argus_render_object_2d_set_transform(self.handle, transform.into()); }
    }

    pub fn copy(&self, parent: RenderGroup2d) -> RenderObject2d {
        unsafe {
            RenderObject2d::of(argus_render_object_2d_copy(self.handle, parent.get_ffi_handle()))
        }
    }
}
